File: server/fastapi/config/routes/39_healthz_app_yaml_stages.routes.py
"""GET /healthz/app-yaml-stage/{stage} — pipeline stage diagnostic accessor."""

from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING, Any

from fastapi import APIRouter, Depends, HTTPException

from app_yaml_fetch_config import (
    get_config,
    get_endpoint,
    list_endpoints,
    resolve_intent,
)

from ._di import get_app_yaml_config, get_app_yaml_loader

if TYPE_CHECKING:
    from ..lifecycles._app_yaml_config import ConfigHandle
    from ..lifecycles._app_yaml_loader import LoaderHandle

VALID_STAGES = ("raw", "merged", "applied", "derived")
EXTRA_CONFIG_FILES = ("database_schema.yaml", "llm_rag.yml", "vite.yaml")


def _stage_raw(loader: "LoaderHandle") -> dict[str, Any]:
    default_loaded = loader.load_from_config_dir(missing="skip")
    extra_paths = [str(Path(loader.config_dir) / f) for f in EXTRA_CONFIG_FILES]
    extra_loaded = loader.load_files(extra_paths, missing="skip")
    out: dict[str, Any] = {}
    for k, v in default_loaded.items():
        out[k] = v
    for k, v in extra_loaded.items():
        out[k] = v
    return out


def _stage_derived() -> dict[str, Any]:
    keys = list_endpoints()
    endpoints = {k: get_endpoint(k) for k in keys}
    applied = get_config() or {}
    declared = list((applied.get("intent_mapping") or {}).get("mappings", {}).keys())
    intent_resolutions: dict[str, Any] = {}
    for intent in declared:
        try:
            intent_resolutions[intent] = resolve_intent(intent)
        except Exception:  # noqa: BLE001
            intent_resolutions[intent] = None
    return {
        "endpoint_keys": keys,
        "endpoints": endpoints,
        "intent_resolutions": intent_resolutions,
        "default_intent": (applied.get("intent_mapping") or {}).get("default_intent"),
    }


def mount(_app, _config) -> APIRouter:
    router = APIRouter()

    @router.get("/healthz/app-yaml-stage/{stage}")
    async def healthz_app_yaml_stage(
        stage: str,
        loader: "LoaderHandle" = Depends(get_app_yaml_loader),
        cfg: "ConfigHandle" = Depends(get_app_yaml_config),
    ):
        if stage not in VALID_STAGES:
            raise HTTPException(
                status_code=400,
                detail={
                    "ok": False,
                    "error": f"unknown stage: {stage}",
                    "valid_stages": list(VALID_STAGES),
                },
            )
        try:
            if stage == "raw":
                data = _stage_raw(loader)
            elif stage == "merged":
                data = cfg.raw
            elif stage == "applied":
                data = get_config()
            else:
                data = _stage_derived()
        except Exception as err:  # noqa: BLE001
            raise HTTPException(status_code=500, detail=str(err)) from err

        return {"ok": True, "stage": stage, "data": data}

    return router

File: server/fastify/config/routes/39_healthz_app_yaml_stages.routes.mjs
// Slot 39 — GET /healthz/app-yaml-stage/:stage  (stage in raw|merged|applied|derived).
// Pure-read accessors over the boot-time pipeline state. Never re-merges or
// re-applies — slot 27/28/29 already did that work. Path uses the
// `app-yaml-stage` prefix to avoid colliding with slot 35's
// `/healthz/app-yaml/:intent` route.

import path from "node:path";
import {
  getConfig,
  listEndpoints,
  getEndpoint,
  resolveIntent,
} from "@ployglot/app-yaml-fetch-config";

const VALID_STAGES = Object.freeze(["raw", "merged", "applied", "derived"]);

const EXTRA_CONFIG_FILES = Object.freeze([
  "database_schema.yaml",
  "llm_rag.yml",
  "vite.yaml",
]);

async function stageRaw(loader) {
  const defaultLoaded = await loader.load_from_config_dir(undefined, {
    missing: "skip",
  });
  const extraPaths = EXTRA_CONFIG_FILES.map((f) =>
    path.join(loader.config_dir, f)
  );
  const extraLoaded = await loader.load_files(extraPaths, { missing: "skip" });
  const out = {};
  for (const [k, v] of defaultLoaded.entries()) out[k] = v;
  for (const [k, v] of extraLoaded.entries()) out[k] = v;
  return out;
}

function stageMerged(cfgSdk) {
  return cfgSdk.getAll();
}

function stageApplied() {
  return getConfig();
}

function stageDerived() {
  const keys = listEndpoints();
  const endpoints = Object.fromEntries(keys.map((k) => [k, getEndpoint(k)]));
  const applied = getConfig();
  const declared = Object.keys(applied?.intent_mapping?.mappings ?? {});
  const intent_resolutions = {};
  for (const intent of declared) {
    try {
      intent_resolutions[intent] = resolveIntent(intent);
    } catch {
      intent_resolutions[intent] = null;
    }
  }
  return {
    endpoint_keys: keys,
    endpoints,
    intent_resolutions,
    default_intent: applied?.intent_mapping?.default_intent ?? null,
  };
}

export default async function healthzAppYamlStagesRoutes(fastify, _config) {
  fastify.get("/healthz/app-yaml-stage/:stage", async (request, reply) => {
    const { stage } = request.params;
    if (!VALID_STAGES.includes(stage)) {
      return reply.code(400).send({
        ok: false,
        error: `unknown stage: ${stage}`,
        valid_stages: VALID_STAGES,
      });
    }

    try {
      let data;
      switch (stage) {
        case "raw":
          data = await stageRaw(request.app_yaml_loader);
          break;
        case "merged":
          data = stageMerged(request.app_yaml_config);
          break;
        case "applied":
          data = stageApplied();
          break;
        case "derived":
          data = stageDerived();
          break;
      }
      return { ok: true, stage, data };
    } catch (err) {
      return reply
        .code(500)
        .send({ ok: false, error: String(err.message ?? err) });
    }
  });
}
